mayastor/src/bdev/dev/nvmx/channel.rs

```diff
@@ -0,0 +1,581 @@
+
+        debug!("{} I/O channel successfully reinitialized", ctrlr_name);
+        self.qpair = Some(qpair);
+        0
+    }
+
+    /// Get I/O statistics for channel.
+    #[inline]
+    pub fn get_io_stats_controller(&mut self) -> &mut IoStatsController {
+        &mut self.io_stats_controller
+    }
+}
+pub struct IoStatsController {
+    // Note that for the sake of optimization, all bytes-related I/O stats
+    // (bytes_read, bytes_written and bytes_unmapped) are accounted in
+    // sectors. Translation into bytes occurs only when providing the full
+    // I/O stats to the caller, inside get_io_stats().
+    io_stats: BlockDeviceIoStats,
+    block_size: u64,
+}
+
+/// Top-level wrapper around device I/O statistics.
+impl IoStatsController {
+    fn new(block_size: u64) -> Self {
+        Self {
+            io_stats: BlockDeviceIoStats::default(),
+            block_size,
+        }
+    }
+
+    #[inline]
+    /// Account amount of blocks and I/O operations.
+    pub fn account_block_io(
+        &mut self,
+        op: IoType,
+        num_ops: u64,
+        num_blocks: u64,
+    ) {
+        match op {
+            IoType::Read => {
+                self.io_stats.num_read_ops += num_ops;
+                self.io_stats.bytes_read += num_blocks;
+            }
+            IoType::Write => {
+                self.io_stats.num_write_ops += num_ops;
+                self.io_stats.bytes_written += num_blocks;
+            }
+            IoType::Unmap => {
+                self.io_stats.num_unmap_ops += num_ops;
+                self.io_stats.bytes_unmapped += num_blocks;
+            }
+            _ => {
+                warn!("Unsupported I/O type for I/O statistics: {:?}", op);
+            }
+        }
+    }
+
+    /// Get I/O statistics for channel.
+    #[inline]
+    pub fn get_io_stats(&self) -> BlockDeviceIoStats {
+        let mut stats = self.io_stats;
+
+        // Translate sectors into bytes before returning the stats.
+        stats.bytes_read *= self.block_size;
+        stats.bytes_written *= self.block_size;
+        stats.bytes_unmapped *= self.block_size;
+
+        stats
+    }
+}
+
+pub struct NvmeControllerIoChannel(NonNull<spdk_io_channel>);
+
+extern "C" fn disconnected_qpair_cb(
+    qpair: *mut spdk_nvme_qpair,
+    _ctx: *mut c_void,
+) {
+    warn!(?qpair, "NVMe qpair disconnected");
+    /*
+     * Currently, just try to reconnect indefinitely. If we are doing a
+     * reset, the reset will reconnect a qpair and we will stop getting a
+     * callback for this one.
+     */
+    unsafe {
+        spdk_nvme_ctrlr_reconnect_io_qpair(qpair);
+    }
+}
+
+extern "C" fn nvme_poll(ctx: *mut c_void) -> i32 {
+    let inner = NvmeIoChannel::from_raw(ctx).inner_mut();
+
+    let num_completions = unsafe {
+        spdk_nvme_poll_group_process_completions(
+            inner.poll_group.as_ptr(),
+            0,
+            Some(disconnected_qpair_cb),
+        )
+    };
+
+    if num_completions > 0 {
+        1
```

Comment:
In here and the other functions below, should we return the spdk enum's or something similar rather than just int literals? Might make it a bit more readable to understand what's being returned

Reply:
Now it uses named constants from the poller module explicitly:
```
if rc == 0 {
    poller::POLLER_IDLE
} else {
   poller::POLLER_BUSY
}
```

Reply:
nice :+1: